shared/datastore.go

```diff
@@ -146,3 +146,60 @@ func VersionPrefix(query *datastore.Query, fieldName, versionPrefix string, desc
 		Filter(fieldName+" >=", fmt.Sprintf("%s.", versionPrefix)).
 		Filter(fieldName+" <=", fmt.Sprintf("%s.%c", versionPrefix, '9'+1))
 }
+
+// GetCompleteRunSHAs returns an array of the SHA[0:10] for runs that
+// exists for all initially-loaded browser names (see GetDefaultBrowserNames),
+// ordered by most-recent.
+func GetCompleteRunSHAs(ctx context.Context, from, to *time.Time, limit *int) (shas []string, err error) {
+	query := datastore.
+		NewQuery("TestRun").
+		Order("-TimeStart")
+
+	// TODO(lukebjerring): Pass in products.
+	products := GetDefaultProducts()
+
+	if from != nil {
+		query = query.Filter("TimeStart >=", *from)
+	}
+	if to != nil {
+		query = query.Filter("TimeStart <", *to)
+	}
+
+	bySHA := make(map[string]mapset.Set)
+	done := mapset.NewSet()
+	it := query.Run(ctx)
+	for {
+		var testRun TestRun
+		var matchingProduct *ProductSpec
+		_, err := it.Next(&testRun)
+		if err == datastore.Done {
+			break
+		} else if err != nil {
+			return nil, err
+		} else {
+			for _, product := range products {
+				if product.Matches(testRun) {
+					matchingProduct = &product
+					break
+				}
+			}
+		}
+		if matchingProduct == nil {
+			continue
+		}
+		set, ok := bySHA[testRun.Revision]
+		if !ok {
+			bySHA[testRun.Revision] = mapset.NewSetWith(*matchingProduct)
```

Comment:
I think we can simply store pointers in the mapset. Not sure if it'd be cheaper.